nabs/plans.py

```diff
@@ -0,0 +1,601 @@
+"""
+Standalone bluesky plans for data collection.
+
+This is the LCLS counterpart to bluesky.plans.
+
+All plans in this module will work as-is when passed into a
+bluesky Run Engine, including starting and stopping a run.
+
+Plans preceded by "daq" incorporate running the LCLS DAQ in the plan.
+"""
+import math
+import time
+from collections import defaultdict
+from itertools import chain, cycle
+
+import bluesky.plan_stubs as bps
+import bluesky.plans as bp
+import bluesky.preprocessors as bpp
+from bluesky import plan_patterns
+from toolz import partition
+
+from . import preprocessors as nbpp
+
+
+def duration_scan(detectors, *args, duration=0, per_step=None, md=None):
+    """
+    Bluesky plan that moves motors among points for a fixed duration.
+
+    For each motor, a list of points must be provided. Each motor will be moved
+    through its list of points simultaneously if multiple motors are provided.
+
+    This will take a reading at every scan step by default via
+    trigger_and_read.
+
+
+    Parameters
+    ----------
+    detectors : list of readables
+        Objects to read into Python in the scan.
+
+    *args :
+        For one dimension, ``motor, [point1, point2, ....]``.
+        In general:
+
+        .. code-block:: python
+
+            motor1, [point1, point2, ...],
+            motor2, [point1, point2, ...],
+            ...,
+            motorN, [point1, point2, ...]
+
+        Motors can be any 'settable' object (motor, temp controller, etc.)
+
+    duration : float
+        The time to run in seconds.
+
+    per_step : plan, optional
+        An alternate plan to run for every scan point. Defaults to
+        one_nd_step.
+
+    md : dict, optional
+        Additional metadata to include in the bluesky start document.
+    """
+    if per_step is None:
+        per_step = bps.one_nd_step
+
+    pos_lists = {}
+    pos_cycles = {}
+    motors = []
+    detectors = list(detectors)
+
+    for motor, pos_list in partition(2, args):
+        pos_list = list(pos_list)
+        pos_lists[motor] = pos_list
+        pos_cycles[motor] = cycle(pos_list)
+        motors.append(motor)
+
+    # md handling lifted from bluesky.plans.list_scan for consistency
+    md_args = list(chain(*((repr(motor), pos_list)
+                           for motor, pos_list in pos_lists.items())))
+
+    _md = {'detectors': [det.name for det in detectors],
+           'motors': [mot.name for mot in motors],
+           'plan_args': {'detectors': list(map(repr, detectors)),
+                         'args': md_args,
+                         'duration': duration,
+                         'per_step': repr(per_step)},
+           'plan_name': 'duration_scan',
+           'plan_pattern': 'inner_list_product',
+           'plan_pattern_module': plan_patterns.__name__,
+           'plan_pattern_args': dict(args=md_args),
+           'hints': {},
+           }
+
+    _md.update(md or {})
+
+    @bpp.stage_decorator(detectors + motors)
```

Comment:
My problem with re-using the built-in scans here is that they are all built with some sort of "move through finite, pre-defined steps" idea, whereas I wanted a scan that had a definite "time" scope rather than a definite "number of steps" scope.

Here I opted to make it as similar to the built-ins as I could, leaving `reset_positions_decorator` for the callers (somewhat ignoring that for LCLS we basically always want reset positions- a mistake?)

Reply:
The implementation is justified by the functionality intent, agreed. It'd only be more complicated to re-use bluesky builtins.

I expect we'll eventually get a complaint that we didn't `reset_positions` for this but pretty much every other - it'd probably be easier to throw it in now than make an API change down the line.

Reply:
Seems reasonable, I'll add it in.